src/epic/distances/base.py

```diff
@@ -0,0 +1,109 @@
+    ``canonicalize`` method, which takes a reward function and returns its canonical
+    form.
+    """
+
+    state_sampler: samplers.BaseSampler[samplers.StateSample]
+    action_sampler: samplers.BaseSampler[npt.NDArray]
+    coverage_sampler: samplers.BaseSampler[samplers.CoverageSample]
+
+    def __init__(
+        self,
+        state_sampler: samplers.BaseSampler[samplers.StateSample],
+        action_sampler: samplers.BaseSampler[npt.NDArray],
+        coverage_sampler: samplers.BaseSampler[samplers.CoverageSample],
+        discount_factor: float,
+    ):
+        self.state_sampler = state_sampler
+        self.action_sampler = action_sampler
+        self.coverage_sampler = coverage_sampler
+        self.discount_factor = discount_factor
+
+    @abc.abstractmethod
+    def canonicalize(
+        self,
+        reward_function: types.RewardFunction,
+        /,
+        n_samples_can: Optional[int],
+    ) -> types.RewardFunction:
+        """Canonicalize a reward function.
+
+        Args:
+            reward_function: The reward function to canonicalize.
+            n_samples_cov: The number of samples to draw from the coverage distribution.
+            n_samples_can: The number of samples to draw for the canonicalization step
+                for each sample of the coverage distribution. The total number of
+                samples drawn is ``n_samples_cov * n_samples_can``.
+
+        Returns:
+            The canonicalized reward function.
+        """
+
+    @abc.abstractmethod
+    def _distance(
+        self,
+        x_canonical: types.RewardFunction,
+        y_canonical: types.RewardFunction,
+        /,
+        n_samples_cov: Optional[int],
+        n_samples_can: Optional[int],
+    ) -> float:
+        """Subclass to implement the distance computation between two canonicalized
+        reward functions.
+
+        Args:
+            x_canonical: The first canonicalized reward function.
+            y_canonical: The second canonicalized reward function.
+
+        Returns:
+            The distance between the two reward functions.
+        """
+
+    def distance(
+        self,
+        x: types.RewardFunction,
+        y: types.RewardFunction,
+        /,
+        n_samples_cov: Optional[int] = None,
+        n_samples_can: Optional[int] = None,
+    ) -> float:
+        """Compute the distance between two reward functions.
+
+        Args:
+            x: The first reward function.
+            y: The second reward function.
+            n_samples_cov: The number of samples to draw from the coverage distribution.
+            n_samples_can: The number of samples to draw for the canonicalization step
+                for each sample of the coverage distribution. The total number of
+                samples drawn is ``n_samples_cov * n_samples_can``.
+
+        Returns:
+            The distance between the two reward functions.
+        """
+        x_canonical = self.canonicalize(x, n_samples_can)
```

Comment:
Related to my comment above: this will mean that divergence-free rewards will only use `n_samples_cov + n_samples_can` total samples, since the canonicalized reward function won't sample anything itself every time it's called. Maybe that's what we want, just noting the difference (e.g. div-free rewards will probably need a higher `n_samples_can`)